Hermes/appscale/hermes/stats/subscribers/cache.py

```diff
@@ -0,0 +1,153 @@
+import threading
+import time
+from datetime import datetime
+
+from appscale.hermes.stats.pubsub_base import StatsSubscriber
+
+BIG_BANG_TIMESTAMP = 0
+
+
+class StatsCache(StatsSubscriber):
+  """
+  It takes care about storing snapshots in limited cache and provides
+  reading method with acknowledgment mechanism.
+  Each node has local stats cache for each kind of stats it collects
+  (node stats, processes stats and haproxy stats for LB nodes).
+  It is used as a temporary storage for stats which wasn't read by master yet.
+  """
+
+  def __init__(self, snapshots_cache_size, ttl=None):
+    self._snapshots_cache = []
+    self._cache_lock = threading.Lock()
```

Comment:
You shouldn't need a threading lock in async code as long as you haven't manually created other threads that operate on the same resources. If you are worried about a yield causing ordering issues, you can use `tornado.locks.Lock`. Since a threading lock doesn't yield, it's easy to end up with a deadlock if you use a threading lock in async code.